Migrate Mcp23xxx from .NET IoT to .NET nanoFramework

mork2020 opened the pull request:
<!--- In the TITLE (↑↑↑↑ above ↑↑↑↑ **NOT HERE**) provide a general, short summary of your changes -->
<!--- Please DO NOT use references to other PR's or issues -->

## Description
<!--- Describe your changes in detail -->
<!--- Bulleted list. Full sentences. Ending with a dot. -->
- migrated Mcp23xxx from .NET IoT to .NET nanoFramework   -   I2C devices only

Here are some file, which are not in the correct "place"
- SpanPinValuePair.cs --> where should this file go? System.Device.Gpio ? 
- PinVector32 --> where should this file go? System.Device.Gpio ?
- PinValuePairEnum.cs --> where should this file go? System.Device.Gpio ?
- GpioDriver.cs --> where should this file go? System.Device.Gpio ?

This file exist because of my error with symbol reference resolving (System.Device.Gpio.PinValuePair)
- PinValuePair.cs

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If this **fixes** OR **closes** OR  **resolves** an open issue, please link to the issue there using the template bellow (mind the pattern to link there as all issues are tracked in the Home repository) -->
<!--- **JUST** replace NNNNN with the issue number -->
- new sensor is supported

## How Has This Been Tested?<!-- (IF APPLICABLE) -->
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- tested with an ESP32 and a MCP23017 breakout board 

## Screenshots<!-- (IF APPROPRIATE): -->

## Types of changes
<!--- What types of changes does this PR introduce? Put an `x` in all the boxes that apply: -->
- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- PLEASE PLEASE PLEASE don't tick all of them just because -->
- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

Ellerbach commented:
> * SpanPinValuePair.cs --> where should this file go? System.Device.Gpio ?

All the Span files are automatically generated from Span<something> into SpanSomething. I don't think you need it as I think we've removed those from the GpioController. So align the functions on the nano System.Device.Gpio. Don't add those if not in ano. That will simplify the code.

> * PinVector32 --> where should this file go? System.Device.Gpio ?

No, all those should stay there. They should not be included in System.Device.Gpio at all. Especially if you align the API on it, I think a large part of those will just go away.

Ellerbach commented on devices/Mcp23xxx/I2cAdapter.cs:
```diff
@@ -0,0 +1,48 @@
+                output[0] = registerAddress;
+                // SpanByte slice is different to Span<Byte> slice in case that index argument (here 1) is equal to size of slice
+                if (data.Length > 0)
+                    data.CopyTo(output.Slice(1));       // do not override output[0]                
```
please use the full braket { and }

mork2020 replied:
corrected

Ellerbach commented on devices/Mcp23xxx/Mcp23xxx.cs:
```diff
@@ -0,0 +1,572 @@
+        /// <inheritdoc/>
+        protected override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => pinNumber;
+
+        /*
```
if you want to keep this one commented, please add a remark why

mork2020 replied:
removed

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Iot.Device.Mcp23xxx
+{
+    public class PinValuePairEnum : IEnumerator
```
if this class is public, please adjust it to add the intellisense comments ///
It has to be on everything public or internal

mork2020 replied:
file not necessary, removed from Mcp23xxx project

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        public Iot.Device.Mcp23xxx.PinValuePair[] _pair;
+        int currentIndex = -1;
+        public PinValuePairEnum(Iot.Device.Mcp23xxx.PinValuePair[] list)
```
Also please respect the coding guidelines with extra line between functions

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            _pair = list;
+        }
+        public bool MoveNext()
```
here as well missing /// and extra line

mork2020 replied:
file not necessary, removed from Mcp23xxx project

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+            currentIndex++;
+            return (currentIndex < _pair.Length);
+        }
+        object IEnumerator.Current
```
here as well and missing an explicit visibility

mork2020 replied:
file not necessary, removed from Mcp23xxx project

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+    public class PinValuePairEnum : IEnumerator
+    {
+        public Iot.Device.Mcp23xxx.PinValuePair[] _pair;
+        int currentIndex = -1;
```
we are using explicit visibility, I guess this one is private. Also for the naming, anything private variable should start with _

mork2020 replied:
file not necessary, removed from Mcp23xxx project

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+                return Current;
+            }
+        }
+        public Iot.Device.Mcp23xxx.PinValuePair Current
```
missing /// and extra line

mork2020 replied:
file not necessary, removed from Mcp23xxx project

Ellerbach commented on devices/Mcp23xxx/PinValuePairEnum.cs:
```diff
@@ -0,0 +1,45 @@
+                }
+            }
+        }
+        public void Reset()
```
missing /// and extra line

mork2020 replied:
file not necessary, removed from Mcp23xxx project

mork2020 commented on devices/Mcp23xxx/GpioDriver.cs:
```diff
@@ -0,0 +1,110 @@
+        /// <param name="value">The value to be written to the pin.</param>
+        protected abstract void Write(int pinNumber, PinValue value);
+
+ 
```
correted

mork2020 commented:
> > * SpanPinValuePair.cs --> where should this file go? System.Device.Gpio ?
> 
> All the Span files are automatically generated from Span into SpanSomething. I don't think you need it as I think we've removed those from the GpioController. So align the functions on the nano System.Device.Gpio. Don't add those if not in ano. That will simplify the code.

I removed the SpanPinValuePair.cs file from the Mcp23xxx project. How is the SpanPinValuePair.cs file generated? The automatic Azure build stops with error that the class is missing.

Ellerbach commented on devices/Mcp23xxx/PinVector32.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown if <paramref name="pinValues"/> contains negative pin numbers or pin numbers higher than 31.
+        /// </exception>
+        public PinVector32(SpanPinValuePair pinValues)
```
Here you have the SpanPinVlauePair. So you may want to simplify this as well

mork2020 replied:
replaced SpanPinValuePair by PinValuePair[]

Ellerbach commented:
> I removed the SpanPinValuePair.cs file from the Mcp23xxx project. How is the SpanPinValuePair.cs file generated? The automatic Azure build stops with error that the class is missing.

it's because you have one left in the code :-) See my comment on this

Ellerbach reviewed (approved):
Thanks a lot @mork2020 for this port! Looks great now!
